ProcessLib/Output.cpp

```diff
@@ -0,0 +1,147 @@
+#include<vector>
+
+#include <logog/include/logog.hpp>
+
+#include "BaseLib/FileTools.h"
+
+namespace
+{
+//! Determines if there should be output at the given \c timestep.
+template<typename CountsSteps>
+bool shallDoOutput(std::size_t timestep, CountsSteps const& repeats_each_steps)
+{
+    std::size_t each_steps = 1;
+
+    for (auto const& pair : repeats_each_steps)
+    {
+        each_steps = pair.each_steps;
+
+        if (timestep > pair.repeat * each_steps)
+        {
+            timestep -= pair.repeat * each_steps;
+        } else {
+            break;
+        }
+    }
+
+    return timestep % each_steps == 0;
+}
+}
+
+namespace ProcessLib
+{
+
+template<typename GlobalSetup>
+std::unique_ptr<Output<GlobalSetup>> Output<GlobalSetup>::
+newInstance(const BaseLib::ConfigTree &config, std::string const& output_directory)
+{
+    std::unique_ptr<Output> out{ new Output{
+        BaseLib::joinPaths(output_directory,
+                           config.getConfParam<std::string>("prefix"))}};
+
+    if (auto const timesteps = config.getConfSubtreeOptional("timesteps"))
+    {
+        for (auto pair : timesteps->getConfSubtreeList("pair"))
+        {
+            auto repeat      = pair.getConfParam<std::size_t>("repeat");
+            auto each_steps = pair.getConfParam<std::size_t>("each_steps");
+
+            assert(repeat != 0 && each_steps != 0);
+            out->_repeats_each_steps.emplace_back(repeat, each_steps);
+        }
+
+        if (out->_repeats_each_steps.empty()) {
+            ERR("You have not given any pair (<repeat/>, <each_steps/>) that defines"
+                    " at which timesteps output shall be written. Aborting.");
+            std::abort();
+        }
+    }
+    else
+    {
+        out->_repeats_each_steps.emplace_back(1, 1);
+    }
+
+    return out;
+}
+
+template<typename GlobalSetup>
+void Output<GlobalSetup>::
+init(typename Output<GlobalSetup>::ProcessIter first,
+     typename Output<GlobalSetup>::ProcessIter last)
+{
+    for (unsigned pcs_idx = 0; first != last; ++first, ++pcs_idx)
+    {
+        auto const filename = _output_file_prefix
+                              + "_pcs_" + std::to_string(pcs_idx)
+                              + ".pvd";
+        _single_process_data.emplace(std::piecewise_construct,
+                std::forward_as_tuple(&**first),
+                std::forward_as_tuple(pcs_idx, filename));
+    }
+}
+
+template<typename GlobalSetup>
+void Output<GlobalSetup>::
+doOutputAlways(Process<GlobalSetup> const& process,
+               std::size_t timestep,
+               const double t,
+               typename GlobalSetup::VectorType const& x)
+{
+    auto spd_it = _single_process_data.find(&process);
+    if (spd_it == _single_process_data.end()) {
+        ERR("The given process is not contained in the output configuration."
+            " Aborting.");
+        std::abort();
+    }
+    auto& spd = spd_it->second;
+
+    std::string const output_file_name =
```

Comment:
I changed my mind and got idea that it might be better to remove a time step number from the VTU file name, i.e. only including only time information. There is a problem in ogs5 benchmarking that benchmarks using adaptive time stepping can produce result files with different names because the file name includes a time step number and the number can change when someone changed numerics.